### PR TITLE
Secretstore RPCs + integration

### PR DESCRIPTION
**Pull request by svyatonik**
On top of #5416 

This is PR for RPCs && small changes required for easier secretstore integration.
First change, as suggested by @ngotchac is to use ethcrypto::encrypt/decrypt with known MAC instead of ethcrypto::{encrypt/decrypt}_single_message. This will help with using RPCs like `parity_decryptMessage`.
@ngotchac I've slightly changed your commit && moved DEFAULT_MAC to ethcrypto from bith RPC && SecretStore modules. Hope you're ok with that :)

### CODE REVIEW

**Comment by svyatonik** [comment]
3 new RPC methods:
```
secretstore_encrypt(account, encrypted_key, plain_data);
secretstore_decrypt(account, encrypted_key, encrypted_data);
secretstore_shadowDecrypt(account, decrypted_secret, common_point, decrypt_shadows, encrypted_data);
```
Where:
```
account - unlocked parity account of KeyServer requestor
encrypted_key - document encryption key, encrypted with requestor public key. It is response of KeyServer generate_document_key() && document_key() REST-methods
plain_data - bytes to encrypt
encrypted_data - bytes to decrypt. It is response of Parity secretstore_encrypt RPC-method
decrypted_secret, common_point, decrypt_shadows - parts of KeyServer document_key_shadow() method response
```
So to encrypt document:
1) call KeyServer::generate_document_key() method:
```
curl -X POST http://localhost:8082/0000000000000000000000000000000000000000000000000000000000000000/de12681e0b8f7a428f12a6694a5f7e1324deef3d627744d95d51b862afc13799251831b3611ae436c452b54cdf5c4e78b361a396ae183e8b4c34519e895e623c00/0
```
2) call Parity::secretstore_encrypt() method:
```
curl --data-binary '{"jsonrpc": "2.0", "method": "secretstore_encrypt", "params": ["0x00dfE63B22312ab4329aD0d28CaD8Af987A01932", "0x047155813de218725ceffb7e6c3dbdf5af4273b96de42552d4575f9de32001c4c2337b685288ee3411b54c5c0ace68d890227b6bc2fa80e52edbd2334bec0a45c9efe989eea16ad26ddf592dd5df55403d475a109350275021676624bb7ee06671f679c275ee26173d6f48030d77ae646b5c6a83ebe5e240b81e5286fa6960f5a998f54ee474f9146820a4ad3260d77bb04660dbbc3492deed8c7a92f138175851f1fb5e565d86464c5eacce7766dc2a7f", "0xdeadbeef"], "id":1 }' -H 'content-type: application/json;' http://127.0.0.1:8545/
```

To decrypt document (simple version, but document secret is revealed to one of KeyServers):
1) call KeyServer::document_key() method:
```
curl http://localhost:8082/0000000000000000000000000000000000000000000000000000000000000000/de12681e0b8f7a428f12a6694a5f7e1324deef3d627744d95d51b862afc13799251831b3611ae436c452b54cdf5c4e78b361a396ae183e8b4c34519e895e623c00
```
2) call Parity::secretstore_decrypt() method:
```
curl --data-binary '{"jsonrpc": "2.0", "method": "secretstore_decrypt", "params": ["0x00dfE63B22312ab4329aD0d28CaD8Af987A01932", "0x047155813de218725ceffb7e6c3dbdf5af4273b96de42552d4575f9de32001c4c2337b685288ee3411b54c5c0ace68d890227b6bc2fa80e52edbd2334bec0a45c9efe989eea16ad26ddf592dd5df55403d475a109350275021676624bb7ee06671f679c275ee26173d6f48030d77ae646b5c6a83ebe5e240b81e5286fa6960f5a998f54ee474f9146820a4ad3260d77bb04660dbbc3492deed8c7a92f138175851f1fb5e565d86464c5eacce7766dc2a7f", "0x2ddec1f96229efa2916988d8b2a82a47ef36f71c"], "id":1 }' -H 'content-type: application/json;' http://127.0.0.1:8545/
```

To decrypt document (extended version, document secret is not revealed to any of KeyServer, except for 1-of-1 case):
1) call KeyServer::document_key_shadow() method:
```
curl http://localhost:8082/shadow/0000000000000000000000000000000000000000000000000000000000000000/de12681e0b8f7a428f12a6694a5f7e1324deef3d627744d95d51b862afc13799251831b3611ae436c452b54cdf5c4e78b361a396ae183e8b4c34519e895e623c00
```
2) call Parity::secretstore_shadowDecrypt() method:
```
curl --data-binary '{"jsonrpc": "2.0", "method": "secretstore_shadowDecrypt", "params":["0x00dfE63B22312ab4329aD0d28CaD8Af987A01932","0x843645726384530ffb0c52f175278143b5a93959af7864460f5a4fec9afd1450cfb8aef63dec90657f43f55b13e0a73c7524d4e9a13c051b4e5f1e53f39ecd91","0x07230e34ebfe41337d3ed53b186b3861751f2401ee74b988bba55694e2a6f60c757677e194be2e53c3523cc8548694e636e6acb35c4e8fdc5e29d28679b9b2f3",["0x049ce50bbadb6352574f2c59742f78df83333975cbd5cbb151c6e8628749a33dc1fa93bb6dffae5994e3eb98ae859ed55ee82937538e6adb054d780d1e89ff140f121529eeadb1161562af9d3342db0008919ca280a064305e5a4e518e93279de7a9396fe5136a9658e337e8e276221248c381c5384cd1ad28e5921f46ff058d5fbcf8a388fc881d0dd29421c218d51761"],"0x2ddec1f96229efa2916988d8b2a82a47ef36f71c"], "id": 1}' -H 'content-type: application/json;' http://127.0.0.1:8545/
```

**Comment by ngotchac** [comment]
The `decrypt` secret store RPC call should be available by default, right? Because a client without the secret store enabled should still be able to call this RPC method in order to decrypt a document.

**Comment by svyatonik** [comment]
Currently it is only supported when building with `--features secretsore`. But you're right - it would be better to support this case. Will update PR. Thanks :)

File: rpc/src/v1/tests/mocked/secretstore.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+use std::sync::Arc;
```
Comment (tomusdrw) [inline rpc/src/v1/tests/mocked/secretstore.rs]: License preamble missing.

File: rpc/src/v1/impls/secretstore.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+mod encryption {
+	use ethkey::{Secret, Public};
+	use jsonrpc_core::Error;
+	use ethcore_secretstore;
+	use util::Bytes;
+	use v1::helpers::errors;
+
+	pub fn encrypt_document(key: Vec<u8>, document: Bytes) -> Result<Bytes, Error> {
+		ethcore_secretstore::encrypt_document(key, document)
+			.map_err(|e| errors::encryption_error(e))
```
Comment (tomusdrw) [inline rpc/src/v1/impls/secretstore.rs]: `map_err(errors::encryption_error)` should be enough. (You should also ditch the `_error` suffix)

File: secret_store/src/encryption.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	let encrypted_document_len = encrypted_document.len();
+	if encrypted_document_len < 16 {
+		return Err(Error::Serde("invalid encrypted data".into()));
+	}
+
+	// make document key
+	let key = into_document_key(key)?;
+
+	// use symmetric decryption to decrypt document
+	let iv = encrypted_document.split_off(encrypted_document_len - 16);
```
Comment (tomusdrw) [inline secret_store/src/encryption.rs]: `encrypted_document_len - 16` is repeated 3 times, maybe its worth to assign it to a variable? `const INIT_VEC_LEN: usize = 16` would also be helpful :)

File: secret_store/src/encryption.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	// key is a previously distributely generated Public
+	if key.len() != 64 {
+		return Err(Error::Serde("invalid public key length".into()));
+	}
+
+	// use x coordinate of distributely generated point as encryption key
+	Ok(key[..16].into())
+}
+
+fn initialization_vector() -> [u8; 16] {
```
Comment (tomusdrw) [inline secret_store/src/encryption.rs]: This could use the same constant.

File: secret_store/src/key_server.rs [hunk 5]
```diff
@@ -143,67 +142,121 @@ impl Drop for KeyServerCore {
...
 		loop {
 			if key_servers.iter().all(|ks| ks.cluster().cluster_state().connected.len() == num_nodes - 1) {
 				break;
 			}
-			if time::Instant::now() - start > time::Duration::from_millis(30000) {
-				panic!("connections are not established in 30000ms");
+
+			let old_tried_reconnections = tried_reconnections;
+			let mut fully_connected = true;
+			for key_server in &key_servers {
```
Comment (tomusdrw) [inline secret_store/src/key_server.rs]: Maybe add some tracing/debug statements here to make it easier to debug the reconnection process?

Comment (svyatonik) [inline secret_store/src/key_server.rs]: I could do, but this is a test. To make traces work I must also add trace initialization in every test which uses this function. So maybe it is ok to leave it as is, until something go wrong?

**Comment by keorn** [comment]
The RPC methods should not require an unlocked account. For now its best to make them take the user password.

**Comment by svyatonik** [comment]
@keorn done

File: rpc/src/v1/traits/secretstore.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+use jsonrpc_core::Error;
+
+use v1::types::{H160, H512, Bytes};
+
+build_rpc_trait! {
+	/// Parity-specific rpc interface.
+	pub trait SecretStore {
+		/// Encrypt data with key, received from secret store.
+		/// Arguments: `account`, `key`, `data`.
```
Comment (keorn) [inline rpc/src/v1/traits/secretstore.rs]: RPC has 4 args

File: rpc/src/v1/traits/secretstore.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		#[rpc(name = "secretstore_encrypt")]
+		fn encrypt(&self, H160, String, Bytes, Bytes) -> Result<Bytes, Error>;
+
+		/// Decrypt data with key, received from secret store.
+		/// Arguments: `account`, `key`, `data`.
```
Comment (keorn) [inline rpc/src/v1/traits/secretstore.rs]: same  here

File: rpc/src/v1/traits/secretstore.rs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		#[rpc(name = "secretstore_decrypt")]
+		fn decrypt(&self, H160, String, Bytes, Bytes) -> Result<Bytes, Error>;
+
+		/// Decrypt data with shadow key, received from secret store.
+		/// Arguments: `account`, `decrypted_secret`, `common_point`, `decrypt_shadows`, `data`.
```
Comment (keorn) [inline rpc/src/v1/traits/secretstore.rs]: arg number

**Comment by svyatonik** [comment]
Tests fail seems unrelated, because only docs were changed in @keorn commit, but it fails, while previous - does not.

**Comment by gavofyork** [comment]
conflicts?

**Pull request merged**